hazelcast/src/main/java/com/hazelcast/query/impl/predicates/InPredicate.java

```diff
@@ -68,15 +69,32 @@ public Predicate accept(Visitor visitor, Indexes indexes) {
 
     @Override
     protected boolean applyForSingleAttributeValue(Comparable attributeValue) {
-        if (attributeValue == null) {
+        Set<Comparable> set = convertedInValues;
+
+        if (attributeValue == null && set == null) {
+            // Conversion of the values given to the predicate is possible only
+            // if the passed attribute value is non-null, otherwise we are
+            // unable to infer a proper converter. Postpone the conversion and
+            // do a brute-force comparison.
+
+            for (Comparable value : values) {
```

Comment:
MInor comment: the presence of `null` values could be cached in a separate `Boolean` variable so that "brute force" is needed at most once.

Reply:
Fixed in https://github.com/hazelcast/hazelcast/pull/16568/commits/ba8d175a844082f6ca40c8c4391e8caf156bb54d